pg-core/src/clj/pg/connection_uri.clj

```diff
@@ -0,0 +1,145 @@
+(ns pg.connection-uri
+  (:require
+   [clojure.string :as str])
+  (:import
+   (java.net URI)))
+
+(set! *warn-on-reflection* true)
+
+(defmacro error! [template & args]
+  `(throw (new Exception (format ~template ~@args))))
+
+(defn parse-query-bool ^Boolean [^String line]
+  (case (str/lower-case line)
+    "" nil
+    ("true" "on" "1" "yes") true
+    ("false" "off" "0" "no") false
+    (error! "cannot parse boolean value: %s" line)))
+
+(defn parse-query-long ^Long [^String line]
+  (case line
+    "" nil
+    (try
+      (Long/parseLong line)
+      (catch Exception e
+        (error! "cannot parse long value: %s, reason: %s"
+          line (ex-message e))))))
+
```

Comment:
This function does most of the things. I thing we only need to extend it with query options.